{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyOwExAIoeldL99mWfwzdUYN"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","execution_count":1,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":305},"id":"zyt9asHFrCU2","executionInfo":{"status":"error","timestamp":1713787637392,"user_tz":-540,"elapsed":7,"user":{"displayName":"Eunsoo Im","userId":"01287249158288146617"}},"outputId":"ebdbb5cd-078e-4493-d51d-87a31cb1288b"},"outputs":[{"output_type":"error","ename":"RuntimeError","evalue":"asyncio.run() cannot be called from a running event loop","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)","\u001b[0;32m<ipython-input-1-230110669ca9>\u001b[0m in \u001b[0;36m<cell line: 72>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     70\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     71\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 72\u001b[0;31m \u001b[0masyncio\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprocess_async\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;32m/usr/lib/python3.10/asyncio/runners.py\u001b[0m in \u001b[0;36mrun\u001b[0;34m(main, debug)\u001b[0m\n\u001b[1;32m     31\u001b[0m     \"\"\"\n\u001b[1;32m     32\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mevents\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_running_loop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 33\u001b[0;31m         raise RuntimeError(\n\u001b[0m\u001b[1;32m     34\u001b[0m             \"asyncio.run() cannot be called from a running event loop\")\n\u001b[1;32m     35\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mRuntimeError\u001b[0m: asyncio.run() cannot be called from a running event loop"]}],"source":["import time\n","import asyncio\n","\n","\n","def test():\n","    pass\n","\n","\n","def test1():\n","    pass\n","\n","\n","def test2():\n","    pass\n","\n","\n","# async\n","test()\n","test1()\n","test2()\n","\n","\n","async def test1():  # coroutine object\n","    await test2()\n","\n","\n","async def test2():\n","    pass\n","\n","\n","def exe_calculate_sync(name, n):\n","    for i in range(i, n + 1):\n","        print(name, i, n)\n","        time.sleep(1)\n","    print(name)\n","\n","\n","def process_sync():\n","    start = time.time()\n","    exe_calculate_sync(\"one\", 1)\n","    exe_calculate_sync(\"two\", 2)\n","    exe_calculate_sync(\"three\", 3)\n","    end = time.time()\n","    print(end - start)\n","\n","\n","async def exe_calculate_async(name, n):\n","    for i in range(i, n + 1):\n","        print(name, i, n)\n","        await asyncio.sleep(1)\n","        # time.sleep(1)\n","    print(name)\n","\n","\n","async def process_async():\n","    start = time.time()\n","\n","    await asyncio.wait(\n","        [\n","            exe_calculate_async(\"one\", 1),\n","            exe_calculate_async(\"two\", 2),\n","            exe_calculate_async(\"three\", 3),\n","        ]\n","    )\n","\n","    end = time.time()\n","    print(end - start)\n","\n","\n","\n","\n","asyncio.run(process_async())\n"]}]}